README.md
# Mixpanel

[![Module Version](https://img.shields.io/hexpm/v/mixpanel_api_ex.svg)](https://hex.pm/packages/mixpanel_api_ex)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/mixpanel_api_ex/)
[![Total Downloads](https://img.shields.io/hexpm/dt/mixpanel_api_ex.svg)](https://hex.pm/packages/mixpanel_api_ex)
[![License](https://img.shields.io/hexpm/l/mixpanel_api_ex.svg)](https://github.com/asakura/mixpanel_api_ex/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/asakura/mixpanel_api_ex.svg)](https://github.com/asakura/mixpanel_api_ex/commits/master)
[![Coverage Status](https://coveralls.io/repos/github/asakura/mixpanel_api_ex/badge.svg?branch=master)](https://coveralls.io/github/asakura/mixpanel_api_ex?branch=master)
[![CI](https://github.com/asakura/mixpanel_api_ex/actions/workflows/elixir.yml/badge.svg)](https://github.com/asakura/mixpanel_api_ex/actions)

This is a non-official third-party Elixir client for the
[Mixpanel](https://mixpanel.com/).

> Note that this README refers to the `master` branch of `mixpanel_api_ex`, not
  the latest released version on Hex. See
  [the documentation](https://hexdocs.pm/mixpanel_api_ex) for the documentation
  of the version you're using.

For the list of changes, checkout the latest
[release notes](https://github.com/asakura/mixpanel_api_ex/CHANGESET.md).

## Installation

Add `mixpanel_api_ex` to your list of dependencies in `mix.exs`.

```elixir
def deps do
  [
    {:mixpanel_api_ex, "~> 1.2"},

    # optional, but recommended adapter
    {:hackney, "~> 1.20"}
  ]
end
```

> The default adapter is Erlang's built-in `httpc`, but it is not recommended to
  use it in a production environment as it does not validate SSL certificates
  among other issues.

And ensure that `mixpanel_api_ex` is started before your application:

```elixir
def application do
  [applications: [:mixpanel_api_ex, :my_app]]
end
```

## Usage Example

Define an interface module with `use Mixpanel`.

```elixir
# lib/my_app/mixpanel.ex

defmodule MyApp.Mixpanel do
  use Mixpanel
end
```

Configure the interface module in `config/config.exs`.

```elixir
# config/config.exs

config :mixpanel_api_ex, MyApp.Mixpanel,
  project_token: System.get_env("MIXPANEL_PROJECT_TOKEN")
```

And then to track an event:

```elixir
MyApp.Mixpanel.track("Signed up", %{"Referred By" => "friend"}, distinct_id: "13793")
# => :ok
```

## TOC

- [Configuration](#configuration)
- [EU Data Residency](#eu-data-residency)
- [Supported HTTP clients](#supported-http-clients)
- [Running multiple instances](#running-multiple-instances)
- [Running tests](#running-tests)
- [Runtime/Dynamic configuration](#runtime-dynamic-configuration)
- [Telemetry](#telemetry)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)

## Configuration

### EU Data Residency

By default `mixpanel_api_ex` sends data to Mixpanels's US Servers. However,
this can be changes via `:base_url` parameter:

```elixir
# config/config.exs

config :mixpanel_api_ex, MyApp.Mixpanel,
  base_url: "https://api-eu.mixpanel.com",
  project_token: System.get_env("MIXPANEL_PROJECT_TOKEN")
```

`:base_url` is not limited specifically to this URL. So if you need you can
provide an proxy address to route Mixpanel events via.

### Supported HTTP clients

At the moment `httpc` and `hackney` libraries are supported. `:http_adapter`
param can be used to select which HTTP adapter you want to use.

```elixir
# config/config.exs

config :mixpanel_api_ex, MyApp.Mixpanel,
  http_adapter: Mixpanel.HTTP.Hackney,
  project_token: System.get_env("MIXPANEL_PROJECT_TOKEN")
```

> The default adapter is Erlang's built-in `httpc`, but it is not recommended to
  use it in a production environment as it does not validate SSL certificates
  among other issues.

### Running multiple instances

You can configure multiple instances to be used by different applications within
your VM. For instance the following example demonstrates having separate client
which is used specifically to sending data to Mixpanel's EU servers.

```elixir
# config/config.exs

config :mixpanel_api_ex, MyApp.Mixpanel,
  project_token: System.get_env("MIXPANEL_PROJECT_TOKEN")

config :mixpanel_api_ex, MyApp.Mixpanel.EU,
  base_url: "https://api-eu.mixpanel.com",
  project_token: System.get_env("MIXPANEL_EU_PROJECT_TOKEN")
```

```elixir
# lib/my_app/mixpanel.ex

defmodule MyApp.Mixpanel do
  use Mixpanel
end
```

```elixir
# lib/my_app/mixpanel_eu.ex

defmodule MyApp.MixpanelEU do
  use Mixpanel
end
```

### Running tests

Other than not running `mixpanel_api_ex` application in test environment you
have got two other options. Which one you need to use depends on if you want the
client process running or not.

If you prefer the client process to be up and running during the test suite
running you may provide `Mixpanel.HTTP.NoOp` adapter to `:http_adapter` param.
As the adapter's name suggests it won't do any actual work sending data to
Mixpanel, but everything else will be running (including emitting Telemetry's
event).

```elixir
# config/test.exs

config :mixpanel_api_ex, MyApp.Mixpanel,
  project_token: "",
  http_adapter: Mixpanel.HTTP.NoOp
```

The second options would be simply assign `nil` as configuration value. This way
that client won't be started by the application supervisor.

```elixir
# config/test.exs

config :mixpanel_api_ex, MyApp.Mixpanel, nil
```

### Runtime/Dynamic Configuration

In cases when you don't know upfront how many client instances you need and what
project tokens to use (for instance this information is read from a database or
a external configuration file during application startup) you can use
`Mixpanel.start_client/1` and `Mixpanel.terminate_client/1` to manually run and
kill instances when needed.

For instance this would start `MyApp.Mixpanel.US` named client with `"token"` project token:

```elixir
Mixpanel.start_client(Mixpanel.Config.client!(MyApp.Mixpanel.US, [project_token: "token"])
# => {:ok, #PID<0.123.0>}
```

`Mixpanel.Config.client!/2` makes sure that provided parameters are correct.

And when you done with it, this function would stop the client immediately:

```elixir
Mixpanel.terminate_client(MyApp.Mixpanel.US)
# => :ok
```

To make it possible to call this client process you might want to use some macro
magic, which would compile a new module in runtime:

```elixir
ast =
  quote do
    use Mixpanel
  end

Module.create(MyApp.Mixpanel.US, ast, Macro.Env.location(__ENV__))
# => {:module, _module, _bytecode, _exports}
```

If creating a module is not a case, you still can call the client's process
directly (it's a GenServer after all). For instance:

```elixir
Client.track(MyApp.Mixpanel.US, event, properties, opts)
# => :ok
```

## Usage

### Tracking events

Use `Mixpanel.track/3` function to track events:

```elixir
MyApp.Mixpanel.track(
  "Signed up",
  %{"Referred By" => "friend"},
  distinct_id: "13793"
)
# => :ok
```

The time an event occurred and IP address of an user can be provided via opts:

```elixir
MyApp.Mixpanel.track(
  "Level Complete",
  %{"Level Number" => 9},
  distinct_id: "13793",
  time: ~U[2013-01-15 00:00:00Z],
  ip: "203.0.113.9"
)
# => :ok
```

### Tracking profile updates

Use `Mixpanel.engage/3,4` function to track profile updates:

```elixir
MyApp.Mixpanel.engage(
  "13793",
  "$set",
  %{"Address" => "1313 Mockingbird Lane"}
)
# => :ok
```

The time an event occurred and IP address of an user can be provided via opts:

```elixir
MyApp.Mixpanel.engage(
  "13793",
  "$set",
  %{"Birthday" => "1948-01-01"},
  time: ~U[2013-01-15 00:00:00Z],
  ip: "123.123.123.123"
)
# => :ok
```

`Mixpanel.engage/2` works with batches:

```elixir
MyApp.Mixpanel.engage(
  [
    {"13793", "$set", %{"Address" => "1313 Mockingbird Lane"}},
    {"13793", "$set", %{"Birthday" => "1948-01-01"}}
  ],
  ip: "123.123.123.123"
)
# => :ok
```

### Merging two profiles

Use `Mixpanel.create_alias/2` create an alias for a district ID, effectively
merging two profiles:

```elixir
MyApp.Mixpanel.create_alias("13793", "13794")
# => :ok
```

## Telemetry

`mixpanel_api_ex` uses Telemetry to provide instrumentation. See the
`Mixpanel.Telemetry` module for details on specific events.

## Contributing

1. Fork it (https://github.com/asakura/mixpanel_api_ex/fork)
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Test your changes by running unit tests and property based tests
   (`mix t && mix p`)
4. Check that provided changes does not have type errors (`mix dialyzer`)
5. Additionally you might run Gradient to have extra insight into type problems
   (`mix gradient`)
6. Make sure that code is formatted (`mix format`)
7. Run Credo to make sure that there is no code readability/maintainability
   issues (`mix credo --strict`)
8. Commit your changes (`git commit -am 'Add some feature'`)
9. Push to the branch (`git push origin my-new-feature`)
10. Create new Pull Request

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

Copyright (c) 2016-2023 [Mikalai Seva](https://github.com/asakura/)


CHANGELOG.md
# Changelog

## 1.2.4

* Support for Erlang/OTP 22, 23, 24, 25 has been added
* Elixir 1.12, 1.13, 1.14 are now supported
* Elixir 1.16.0-rc.0 is supported on OTP 24, 25, 26 as well
* If a project token present in properties it won't be overwritten
* Dynamic starting/stopping of client processes was implemented

## 1.2.3

* NoOp was added to the list of known adapters
* mix_unused was listed as dev only dependency

## 1.2.2

* Add VERSION to list of the package's extra files

## 1.2.1

* Fixes name of the default HTTP adapter

## 1.2.0

### Breaking changes

* Now you can configure the library in a such a way that it can be used to
  stream to multiple Mixpanel accounts simultaneously. Consult with README.md on
  how to use new API.

### Improvements

* Introduces Telemetry support.

## 1.1.0

### Improvements

* Supports Elixir 1.15 and Erlang/OTP 26: all compilation errors and warnings
  were fixed.
* User facing API now supports NaiveDateTime, DateTime, as well as Erlang's
  timestamps and Erlang's Calendar `t:datetime()`.
* Batch variant of engage function (`Mixpanel.engage/2`) has been added.
* `Mixpanel.create_alias/2` has been added.
* Poison has been replaced with Jason.
* `base_url` option has been added which enables selection EU Residency servers
  (or enables use of proxies of your choice).
* Supports plug-able adapters for http libraries via `:http_adapter`
  configuration parameter: implements default one using `httpc` OTP library and
  and another one using Hackney library (won't be available until it listed as
  dependency in your project's mix file).
* If a request to Mixpanel backend fails for any reason, then it will be retries
  up to 3 times.
* All dependencies have been upgraded to their latest versions.

### Breaking changes

* Mentions of the Token has been replaced with Project Token to reflect the
  official docs. Rename `token` to `project_token` in the `config.exs`.

### Deprecation

* HTTPoison has been removed
* Mock library was made obsolete
* InchEx integration was removed
* Dogma has been dropped

### General code quality improvements

* Travis has been obsolete and was replaced by GitHub actions.
* Now the library uses Dependabot.
* All typespecs were refined and thus improved the documentation.
* Dialyzer errors have been fixed.
* Credo's warnings and errors have been resolved
* A lot of code repetition has been eliminated.
* Better validation of user provided options has been added to user facing API
  functions.
* Since the library employs behaviours to implement HTTP client adapters, the
  test suit was moved to Mox mocking framework and thus was simplified.
* To test HTTP adapters, Bandit running in HTTPS mode was used.
* Due to changes listed above, the test suite coverage was greatly improved.

## 0.8.4

* Update deps

## 0.8.0

Initial release


mix.exs
defmodule Mixpanel.Mixfile do
  use Mix.Project

  @external_resource "VERSION"
  @version File.read!(Path.join([__DIR__, "VERSION"]))

  def project do
    [
      app: :mixpanel_api_ex,
      version: @version,
      elixir: "~> 1.12",
      elixirc_paths: elixirc_paths(Mix.env()),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      compilers: compilers(Mix.env()),
      unused: unused(),
      deps: deps(),
      aliases: aliases(),
      preferred_cli_env: preferred_cli_env(),
      dialyzer: [
        plt_core_path: "priv/plts",
        plt_file: {:no_warn, "priv/plts/dialyzer.plt"},
        plt_add_deps: :apps_direct,
        plt_add_apps: [:logger, :inets],
        plt_ignore_apps: [:mix_unused, :propcheck],
        flags: [
          "-Werror_handling",
          "-Wextra_return",
          "-Wmissing_return",
          "-Wunknown",
          "-Wunmatched_returns",
          "-Wunderspecs"
        ]
      ],
      propcheck: [counter_examples: "propcheck_counter_examples"],
      test_paths: test_paths(Mix.env()),
      test_coverage: [
        tool: ExCoveralls,
        export: "cov",
        test_task: "test",
        summary: [
          threshold: 80
        ]
      ],

      # Hex
      description: description(),
      package: package(),

      # Docs
      name: "Mixpanel API",
      docs: [
        extras: ["README.md", "CHANGELOG.md"],
        source_ref: "v#{@version}",
        main: "Mixpanel",
        source_url: "https://github.com/asakura/mixpanel_api_ex"
      ]
    ]
  end

  def description do
    "Elixir client for the Mixpanel API."
  end

  defp elixirc_paths(:test), do: ["test/support", "lib"]
  defp elixirc_paths(:property), do: ["property/support", "lib"]
  defp elixirc_paths(_), do: ["lib"]

  defp test_paths(:property), do: ["property"]
  defp test_paths(_), do: ["test"]

  def package do
    [
      maintainers: ["Mikalai Seva"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/asakura/mixpanel_api_ex"},
      files: ~w(mix.exs README.md CHANGELOG.md lib VERSION)
    ]
  end

  def application() do
    [
      mod: {Mixpanel, []},
      extra_applications: [:logger],
      start_phases: [
        clients: []
      ]
    ]
  end

  defp compilers(:dev) do
    [:unused] ++ Mix.compilers()
  end

  defp compilers(_), do: Mix.compilers()

  defp deps do
    [
      {:bandit, "~> 1.0", only: :test},
      # only to make excoveralls to work on OTP 24
      {:castore, "~> 1.0", only: :test},
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.4", only: [:dev, :test], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:excoveralls, "~> 0.18", only: [:test, :property]},
      # Erlang 22 comparability layer
      # Remove it when OTP 22 support is dropped
      unquote_splicing(
        if :erlang.system_info(:otp_release) |> to_string() == "22" do
          []
        else
          quote do
            [
              {:gradient, github: "esl/gradient", ref: "33e13fb", only: [:dev], runtime: false}
              # {:gradient_macros, github: "esl/gradient_macros", ref: "3bce214", runtime: false},
            ]
          end
        end
      ),
      {:hackney, "~> 1.20", only: [:test, :dev]},
      {:jason, "~> 1.4"},
      {:machete, "~> 0.2", only: :test},
      {:mix_unused, "~> 0.4.1", only: :dev},
      {:mox, "~> 1.1", only: :test},
      {:propcheck, "~> 1.4", only: [:property, :dev]},
      {:telemetry, "~> 0.4 or ~> 1.0"}
    ]
  end

  defp preferred_cli_env do
    [
      c: :dev,
      t: :test,
      ti: :test,
      p: :property,
      "test.property": :property,
      coveralls: :test,
      "coveralls.github": :test,
      "coveralls.detail": :test,
      "coveralls.post": :test,
      "coveralls.html": :test,
      "coveralls.cobertura": :test
    ]
  end

  defp aliases() do
    [
      c: "compile",
      t: "test --no-start",
      p: &run_property_tests/1,
      d: "dialyzer",
      g: "gradient",
      test: "test --no-start",
      "test.property": &run_property_tests/1
    ]
  end

  defp unused() do
    [
      ignore: [
        {:_, :child_spec, :_},
        {:_, :start_link, :_}
      ]
    ]
  end

  defp run_property_tests(args) do
    env = Mix.env()
    args = if IO.ANSI.enabled?(), do: ["--color" | args], else: ["--no-color" | args]
    IO.puts("Running tests with `MIX_ENV=#{env}`")

    {_, res} =
      System.cmd("mix", ["test" | args],
        into: IO.binstream(:stdio, :line),
        env: [{"MIX_ENV", to_string(env)}]
      )

    if res > 0 do
      System.at_exit(fn _ -> exit({:shutdown, 1}) end)
    end
  end
end
